tt_metal/hw/inc/dataflow_api.h

```diff
@@ -1525,11 +1528,50 @@ void cq_pop_front(uint32_t cmd_size_B) {
     // host and device are consistent in updating their pointers in this way, so they won't get out of sync. The
     // alignment is necessary because we can only read/write from/to 32B aligned addrs in host<->dev communication.
     uint32_t cmd_size_16B = align(cmd_size_B, 32) >> 4;
-    cq_read_interface.fifo_rd_ptr += cmd_size_16B;
+    cq_read_interface.issue_fifo_rd_ptr += cmd_size_16B;
 
     notify_host_of_cq_read_pointer();
 }
 
+FORCE_INLINE
+void cq_reserve_back(uint32_t data_size_B) {
+    DEBUG_STATUS('N', 'Q', 'R', 'B', 'W');
+    uint32_t data_size_16B = align(data_size_B, 32) >> 4;
+    uint32_t rd_ptr_and_toggle;
+    uint32_t rd_ptr;
+    uint32_t rd_toggle;
+    do {
+        rd_ptr_and_toggle = *get_cq_completion_read_ptr();
+        rd_ptr = rd_ptr_and_toggle & 0x7fffffff;
+        rd_toggle = rd_ptr_and_toggle >> 31;
+    } while (
+        cq_write_interface.completion_fifo_wr_ptr < rd_ptr and cq_write_interface.completion_fifo_wr_ptr + data_size_16B > rd_ptr or
+        (rd_toggle != cq_write_interface.completion_fifo_wr_toggle and cq_write_interface.completion_fifo_wr_ptr == rd_ptr)
+    );
+
+    DEBUG_STATUS('N', 'Q', 'R', 'B', 'D');
+}
+
+FORCE_INLINE
+void cq_push_back(uint32_t push_size_B) {
+    uint32_t push_size_16B = align(push_size_B, 32) >> 4;
+    cq_write_interface.completion_fifo_wr_ptr += push_size_16B;
+    if (cq_write_interface.completion_fifo_wr_ptr >= cq_write_interface.completion_fifo_limit) {
+        cq_write_interface.completion_fifo_wr_ptr = DeviceCommand::COMMAND_ISSUE_REGION_SIZE >> 4;
```

Comment:
DeviceCommand::COMMAND_ISSUE_REGION_SIZE should be passed in as a compile-time variable.